File: EngSoftTypeScript/src/Pato.ts
import {PadraoVoaveis} from "./Padro_Voaveis";

export abstract class Pato {

    protected _comportamentoVoar!: PadraoVoaveis;

    abstract mostrar(): string;

    public nadar(): string {
        return "Pato Nadando.";
    }

    set comportamentoVoar(value: PadraoVoaveis) {
        this._comportamentoVoar = value;
    }

    public voar(): string {
        return this._comportamentoVoar.voar();
    }

}

File: EngSoft.Jogo.Patos/src/engsoft/jogo/patos/Pato.java
package engsoft.jogo.patos;

// Classe abstrata
// Programar para supertipo
public abstract class Pato {

	// Programar para supertipo
	protected Padrao_Voaveis comportamento_pato;

	abstract String mostrar();

	// Este método sempre retorna a mesma coisa, logo, não é polimórfico
	public String nadar() {
		return "Pato Nadando.";
	}	
	
	public void setComportamento(Padrao_Voaveis padrao ) {
		comportamento_pato = padrao;
	}
	
	public String comportamento_pato() {
		return comportamento_pato.voar();
	}

}
